### ISSUE TITLE
Graceful handling of 404s which may be auth errors

### ISSUE THREAD

**Issue by StephenWithPH** [integrations/terraform-provider-github#308]
Today, `github_repository_deploy_key` does bad things when Terraform receives a `404` from GitHub because the user-provided GitHub api token doesn't have sufficient permissions to manage repository deploy keys. This can happen when _reading_ the resource's state from GitHub's api upon refresh.

This can arise in environments where multiple users collaborate on the same set of Terraform configs which manage GitHub, but those users have different permissions.

### Terraform Version
```sh
Terraform v0.12.5
+ provider.github v2.2.1
```

### Affected Resource(s)
- github_repository_deploy_key

(there are likely others)

### Terraform Configuration Files
```hcl
provider "github" {
  version = "~> 2.2"

  organization = <redacted>
  token        = <redacted>
}


resource "tls_private_key" "this" {
  algorithm = "RSA"
  rsa_bits  = 2048
}

resource "github_repository_deploy_key" "this" {
  key        = tls_private_key.this.public_key_openssh
  read_only  = true
  repository = <redacted>
  title      = "my-key"
}
```

### Feature Request
Perhaps a property that is spiritually similar to [force_destroy](https://www.terraform.io/docs/providers/aws/r/s3_bucket.html#force_destroy) from other contexts. This would allow users to tell Terraform that once this resource exists in Terraform's state (i.e., it has been successfully created), do _not_ remove it from state upon 404. Instead, fail with a warning of "resource appears gone, but this may be due to permissions. Good luck!"

```hcl
resource "github_repository_deploy_key" "this" {
  key        = tls_private_key.this.public_key_openssh
  read_only  = true
  repository = <redacted>
  title      = "my-key"

  fail_on_missing = true #<-- this is the feature request
}
```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply` the above configurations using a GitHub token with sufficient scope to manage repo deploy keys.
1. `terraform apply` using a different GitHub token... one with insufficient scope.

### Important Factoids
GitHub **deliberately** masks permissions errors with 404s:
>Typically, we send a 404 error when your client isn't properly authenticated. You might expect to see a 403 Forbidden in these cases. However, since we don't want to provide any information about private repositories, the API returns a 404 error instead.

(https://developer.github.com/v3/troubleshooting/#why-am-i-getting-a-404-error-on-a-repository-that-exists)

The behavior in question happens here:
https://github.com/terraform-providers/terraform-provider-github/blob/ca39a4d23637a1b0ff8b7463e93b5e08de0e74e8/github/resource_github_repository_deploy_key.go#L119-L124

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- #62 
- #176 

(These are loosely related in that they seem to be other cases of auth-error-masked-as-`404` causes user pain)

**Comment by majormoses** [comment]
> This can arise in environments where multiple users collaborate on the same set of Terraform configs which manage GitHub, but those users have different permissions.

I would highly encourage you to avoid this all together and use something like jenkins, atlantis, or terraform enterprise to have a system for interacting with github (with terraform) rather than people running things locally on their workstations when making administrative changes.

I am not opposed to the concept (though I probably not thought through all the implications) but I think if implemented it should be on every object within this provider and not cherry pick individual ones. For example for safety we do not allow the access token that our automated systems to delete repositories for safety. Given that repository deletions in my org are in the minority and the cost of accidentally blowing up a repository is high we require a manual repository deletion (making a backup/archive first just in case) from a github admin. Due to this removing objects that are not found from the state make a lot of sense. I think we can probably have our cake and eat it too but it needs further thought/exploration IMO.

**Comment by StephenWithPH** [comment]
> I would highly encourage you to avoid this all together and use something like jenkins, atlantis, or terraform enterprise to have a system for interacting with github (with terraform) rather than people running things locally on their workstations when making administrative changes

In general, that is exactly what we do. In this case, we have a bootstrapping problem: the build agents which will run Terraform in the future first need to be configured with deploy keys, IAM roles, etc. That privileged, infrequent action has to be run by a human. Because it is infrequent yet important, we automated it with Terraform.

We learned that some privileged users were more privileged than others in GitHub when Terraform removed the keys. We've since corrected that GitHub misconfiguration, but we thought we'd raise the issue here for discussion.

> I think if implemented it should be on every object within this provider and not cherry pick individual ones.

I agree.

> I think we can probably have our cake and eat it too but it needs further thought/exploration IMO.

I also agree. This seems like a specific case of drift detection when the resource API reports full absence rather than changes to fields. Due to the behavior of the resource API, we want to give users a way to tell Terraform how to handle a completely missing resource which previously existed. My initial `fail_on_missing` suggestion was meant to spur discussion... it's a little gross, but I can't think of a more elegant approach.
